README.md
# PHP Code Generator

This project aims to deliver a easy to use free as in freedom php code generator.
PHPDOC-Generation is also implemented as general generation tasks like "class", "function" or "property" generation.

The build status of the current master branch is tracked by Travis CI: 
[![Build Status](https://travis-ci.org/bazzline/php_component_code_generator.png?branch=master)](http://travis-ci.org/bazzline/php_component_code_generator)
[![Latest stable](https://img.shields.io/packagist/v/net_bazzline/php_component_code_generator.svg)](https://packagist.org/packages/net_bazzline/php_component_code_generator)

The scrutinizer status are:
[![code quality](https://scrutinizer-ci.com/g/bazzline/php_component_code_generator/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/bazzline/php_component_code_generator/) | [![build status](https://scrutinizer-ci.com/g/bazzline/php_component_locator_generator/badges/build.png?b=master)](https://scrutinizer-ci.com/g/bazzline/php_component_code_generator/)

The versioneye status is:
[![Dependency Status](https://www.versioneye.com/user/projects/54036dbbeab62ac615000143/badge.svg?style=flat)](https://www.versioneye.com/user/projects/54036dbbeab62ac615000143)

This component comes with a lot of [examples](https://github.com/bazzline/php_component_code_generator/tree/master/example) as well as default [factories](https://github.com/bazzline/php_component_code_generator/tree/master/source/Net/Bazzline/Component/CodeGenerator/Factory).

Take a look on [ohloh.net](https://www.ohloh.net/p/php_component_code_generator).

The current change log can be found [here](https://github.com/bazzline/php_component_code_generator/blob/master/CHANGELOG.md).

# Install

## By Hand

```
mkdir -p vendor/net_bazzline/php_component_code_generator
cd vendor/net_bazzline/php_component_code_generator
git clone https://github.com/bazzline/php_component_code_generator .
```

## With [Packagist](https://packagist.org/packages/net_bazzline/php_component_code_generator)

```
composer require net_bazzline/php_component_code_generator:dev-master
```

# Benefits

* no "new" inside classes ...
* independent and configurable indention
* no static calls
* shipped with examples and factories
* covered with unittests
* open source
* automatic phpdoc generation

# Optimization Potential 

* ... but "clone"
* currently not widely used 
* no code review by other developers so far
* still open todo's

# API

[API](http://www.bazzline.net/57b93db867f58f6bf2982833765721032a5ea22b/index.html) is available at [bazzline.net](http://www.bazzline.net).

# Inspired By

* [php-generator](https://github.com/nette/php-generator)
* [simple-php-code-gen](https://github.com/gotohr/simple-php-code-gen)
* [cg-library](https://github.com/schmittjoh/cg-library)
* [sensio generator bundle](https://github.com/sensiolabs/SensioGeneratorBundle)
* [php-ide-stub-generator](https://github.com/racztiborzoltan/php-ide-stub-generator)
* [php-token-reflection](https://github.com/Andrewsville/PHP-Token-Reflection)
* [code-generator](https://github.com/Speicher210/CodeGenerator)

# Final Words

Star it if you like it :-). Add issues if you need it. Pull patches if you enjoy it. Write a blog entry if you use it. [Donate something](https://gratipay.com/~stevleibelt) if you love it :-].


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Open]

### To Add

* add type hint "array" wherever possible and not only in the documentation
* implement possibility to use some ["pre defined" methods](https://github.com/wells5609/CodeGenerator/tree/master/src/CodeGenerator/Method) for generation
* implement method ordering by visibility and final|!final status in ClassGenerator to get some automatic ordering of public|protected|private methods for generation
* implement [annotation](https://github.com/Speicher210/CodeGenerator/tree/master/src/Wingu/OctopusCore/CodeGenerator/PHP/Annotation)
* implement validation
* implement [automatically documentation](https://github.com/wells5609/CodeGenerator) (validate if this is still better than current approach)

### To Change

* [extend trait](https://github.com/Speicher210/CodeGenerator/blob/master/docs/php/oop/generate-trait.md)
* [extend indention](https://github.com/Speicher210/CodeGenerator/blob/master/src/Wingu/OctopusCore/CodeGenerator/GeneratorInterface.php)
* [flavour api](https://github.com/propelorm/Propel/blob/master/generator/lib/builder/om/OMBuilder.php)
* remove @todo's in source code
* use symfony console and commands
* what about something like [that](https://github.com/zetacomponents/PhpGenerator/blob/master/docs/example_general.php)

## [Unreleased]

### Added

### Changed

## [1.2.1](https://github.com/bazzline/php_component_code_generator/tree/1.2.1) - released at 2017-07-26

### Changed

* fixed broken composer.json

## [1.2.0](https://github.com/bazzline/php_component_code_generator/tree/1.2.0) - released at 2017-05-24

### Added

* added api link
* added integration testing for php 7.0

### Changed

* replaced deprecated array syntax with new one
* removed integration testing for php < 5.6
* removed shipped with api documentation

## [1.1.11](https://github.com/bazzline/php_component_code_generator/tree/1.1.11) - released at 2016-02-20

### Changed

* moved to psr-4 autoloading
* updated dependencies

## [1.1.10](https://github.com/bazzline/php_component_code_generator/tree/1.1.10) - released at 2015-012-11

### Changed

* updated dependencies

## [1.1.9](https://github.com/bazzline/php_component_code_generator/tree/1.1.9) - released at 2015-11-08

### Changed

* updated dependencies

## [1.1.8](https://github.com/bazzline/php_component_code_generator/tree/1.1.8) - released at 2015-08-21

### Changed

* updated dependencies

## [1.1.7](https://github.com/bazzline/php_component_code_generator/tree/1.1.7) - released at 2015-07-29

### Changed

* updated dependencies

## [1.1.6](https://github.com/bazzline/php_component_code_generator/tree/1.1.6) - released at 2015-07-04

### Changed

* updated dependencies

## [1.1.5](https://github.com/bazzline/php_component_code_generator/tree/1.1.5) - released at 2015-05-22

### Changed

* updated dependencies

## [1.1.4](https://github.com/bazzline/php_component_code_generator/tree/1.1.4) - released at 2015-02-08

### Changed

* removed dependency to apigen

## [1.1.3](https://github.com/bazzline/php_component_code_generator/tree/1.1.3) - released at 2015-01-07

### Changed

* fixed bug in SignatureGenerator when auto generating Documentation as Interface

## [1.1.2](https://github.com/bazzline/php_component_code_generator/tree/1.1.2) - released at 2015-01-07

### Added

* added support for InterfaceGenerator to DocumentationGenerator

## [1.1.1](https://github.com/bazzline/php_component_code_generator/tree/1.1.1) - released at 2015-01-07

### Added

* added support for InterfaceGenerator to FileGenerator

## [1.1.0](https://github.com/bazzline/php_component_code_generator/tree/1.1.0) - released at 2015-01-06

### Added

* added api
* created a general SignatureGenerator (extended by InterfaceGenerator and ClassGenerator)

### Changed

* fixed a bug in class generator by creating a interface generator
* by using a interface generator, it is now possible to create an interface that extends more than one interface
* covered new code with unit tests
* updated InterfaceExample
* updated dependencies
* [migration](https://github.com/bazzline/php_component_code_generator/blob/master/migration/1.0.1_to_1.1.0.md) needed

## [1.0.1](https://github.com/bazzline/php_component_code_generator/tree/1.0.1) - released at 2014-08-31

### Changed

* transfered project to bazzline
* updated dependencies

## [1.0.0](https://github.com/bazzline/php_component_code_generator/tree/1.0.0) - released at 2014-07-27

### Added

* added example for [interface generator](https://github.com/bazzline/php_component_code_generator/tree/1.0.0/example/InterfaceExample.php)
* added example for [file generator](https://github.com/bazzline/php_component_code_generator/tree/1.0.0/example/FileExample.php)

## [0.0.3](https://github.com/bazzline/php_component_code_generator/tree/0.0.3) - released at 2014-06-10

### Changed

* fixed broken strict tests
* fixed invalid generation of interfaces

## [0.0.2](https://github.com/bazzline/php_component_code_generator/tree/0.0.2) - released at 2014-06-05

### Added

* added method Indention::isSetToInitialLevel();

### Changed

* fixed bug in BlockGenerator::startIndention() - now multiple calls are supported
* fixed logical bug by replacing "addExtends" to "setExtends" since it is not allowed to extend from more than one class

## [0.0.1](https://github.com/bazzline/php_component_code_generator/tree/0.0.1) - released at 2014-05-25

### Added

* covered main code with tests
* created [examples](https://github.com/bazzline/php_component_code_generator/tree/0.0.1/example)
* created [factories](https://github.com/bazzline/php_component_code_generator/tree/0.0.1/source/Net/Bazzline/Component/CodeGenerator/Factory)


source/MethodGenerator.php
<?php
/**
 * @author stev leibelt <artodeto@bazzline.net>
 * @since 2014-04-24 
 */

namespace Net\Bazzline\Component\CodeGenerator;

/**
 * Class MethodGenerator
 * @package Net\Bazzline\Component\Locator\LocatorGenerator\Generator
 */
class MethodGenerator extends AbstractDocumentedGenerator
{
    /**
     * @param string $name
     * @param string $defaultValue
     * @param string $typeHint
     * @param bool $isReference
     * @return $this
     */
    public function addParameter($name, $defaultValue = '', $typeHint = '', $isReference = false)
    {
        $parameter = [
            'default_value' => $defaultValue,
            'name'          => $name,
            'is_reference'  => $isReference,
            'type_hint'     => $typeHint
        ];

        $this->addGeneratorProperty('parameters', $parameter);

        return $this;
    }

    /**
     * @param float|string $version
     * @param string $description
     * @return $this
     * @see http://www.phpdoc.org/docs/latest/for-users/phpdoc/tags/since.html
     */
    public function addSince($version, $description = '')
    {
        $since = [
            'description'   => $description,
            'version'       => $version
        ];

        $this->addGeneratorProperty('since_versions', $since);

        return $this;
    }

    /**
     * @return null|array
     */
    public function getBody()
    {
        return $this->getGeneratorProperty('body', null);
    }

    /**
     * @param string|array|LineGenerator|BlockGenerator $body
     * @param array $returnValueTypeHints
     * @return $this
     */
    public function setBody($body, $returnValueTypeHints = [])
    {
        $this->addGeneratorProperty('body', $body, false);
        $this->addGeneratorProperty('has_body', true, false);
        $this->addGeneratorProperty('body_return_value_type_hints', $returnValueTypeHints, false);

        return $this;
    }

    /**
     * @return $this
     */
    public function markAsAbstract()
    {
        $this->addGeneratorProperty('abstract', true, false);

        return $this;
    }

    /**
     * @return $this
     */
    public function markAsHasNoBody()
    {
        $this->addGeneratorProperty('has_body', false, false);

        return $this;
    }

    /**
     * @return $this
     */
    public function markAsFinal()
    {
        $this->addGeneratorProperty('final', true, false);

        return $this;
    }

    /**
     * @return $this
     */
    public function markAsPrivate()
    {
        $this->addGeneratorProperty('visibility', 'private', false);

        return $this;
    }

    /**
     * @return $this
     */
    public function markAsProtected()
    {
        $this->addGeneratorProperty('visibility', 'protected', false);

        return $this;
    }

    /**
     * @return $this
     */
    public function markAsPublic()
    {
        $this->addGeneratorProperty('visibility', 'public', false);

        return $this;
    }

    /**
     * @return $this
     */
    public function markAsStatic()
    {
        $this->addGeneratorProperty('static', true, false);

        return $this;
    }

    /**
     * @param string $name
     * @return $this
     */
    public function setName($name)
    {
        $this->addGeneratorProperty('name', (string) $name, false);

        return $this;
    }

    /**
     * @throws InvalidArgumentException|RuntimeException
     * @return string
     * @todo implement exception throwing if mandatory parameter is missing
     */
    public function generate()
    {
        $this->resetContent();
        $this->generateDocumentation();
        $this->generateSignature();
        $this->generateBody();

        return $this->generateStringFromContent();
    }

    private function generateBody()
    {
        $body       = $this->getGeneratorProperty('body', ['//@todo implement']);
        $isAbstract = $this->getGeneratorProperty('abstract', false);
        $hasBody    = $this->getGeneratorProperty('has_body', true);

        if (!$isAbstract
            && $hasBody) {
            $this->addContent('{');
            if (!($body instanceof BlockGenerator)
                || !($body instanceof LineGenerator)) {
                if (!is_array($body)) {
                    $body = [
                        $body
                    ];
                }
                $body = $this->getBlockGenerator($body);
            }
            $this->addGeneratorAsContent(
                $body,
                true
            );
            $this->addContent('}');
        }
    }

    private function generateDocumentation()
    {
        $documentation = $this->getGeneratorProperty('documentation');

        if ($documentation instanceof DocumentationGenerator) {
            if ($this->completeDocumentationAutomatically === true) {
                $parameters         = $this->getGeneratorProperty('parameters', []);
                $returnTypeHints    = $this->getGeneratorProperty('body_return_value_type_hints');

                foreach ($parameters as $parameter) {
                    $documentation->addParameter($parameter['name'], $parameter['type_hint']);
                }
                if (is_array($returnTypeHints)) {
                    $documentation->setReturn($returnTypeHints);
                }
            }
            $this->addGeneratorAsContent($documentation);
        }
    }

    private function generateSignature()
    {
        if ($this->canBeGenerated()) {
            $hasBody        = $this->getGeneratorProperty('has_body', true);
            $isAbstract     = $this->getGeneratorProperty('abstract', false);
            $isFinal        = $this->getGeneratorProperty('final', false);
            $isStatic       = $this->getGeneratorProperty('static', false);
            $name           = $this->getGeneratorProperty('name');
            $parameters     = $this->getGeneratorProperty('parameters', []);
            $visibility     = $this->getGeneratorProperty('visibility');

            //@todo refactor the wired usage for line and block generator
            $line = $this->getLineGenerator();

            if ($isAbstract) {
                $line->add('abstract');
            } else if ($isFinal) {
                $line->add('final');
            }

            if (!is_null($visibility)) {
                $line->add($visibility);
            }

            if ($isStatic) {
                $line->add('static');
            }

            $parametersLine = $this->getLineGenerator();
            $parametersLine->setSeparator(', ');
            if (is_array($parameters)) {
                foreach ($parameters as $parameter) {
                    $parameterLine = $this->getLineGenerator();
                    if ((strlen($parameter['type_hint']) > 0)
                        && (!in_array($parameter['type_hint'], $this->getNotPrintableTypeHints()))) {
                        $parameterLine->add($parameter['type_hint']);
                    }
                    $parameterLine->add(($parameter['is_reference'] ? '&' : '') . '$' . $parameter['name']);
                    if (strlen((string) $parameter['default_value']) > 0) {
                        $parameterLine->add('= ' . (string) $parameter['default_value']);
                    }
                    $parametersLine->add($parameterLine);
                }
            }

            $line->add('function ' . $name . '(' . $parametersLine->generate() . ')' . ((($isAbstract) || (!$hasBody)) ? ';' : ''));
            $block = $this->getBlockGenerator($line);
            $this->addContent($block);
        }
    }
}
